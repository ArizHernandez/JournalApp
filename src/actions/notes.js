import Swal from 'sweetalert2';

import { db } from '../firebase/firebase-config';
import { fileUpload } from '../helpers/fileUpload';
import { loadNotes } from '../helpers/loadNotes';
import { Types } from '../types/types';

export const startNewNote = () => {
  return async(dispatch, getState) => {
    const {uid} = getState().auth
    const newNote = {
      title: '',
      body:  '',
      date:  new Date().getTime(),
      imageUrl: ''
    }

    const docRef = await db.collection(`${uid}/journal/notes`).add(newNote);

    dispatch(activeNote(docRef.id, newNote));
    dispatch(addNewNote(docRef.id, newNote));
  }
};

export const startLoadingNotes = (uid) => {
  return async(dispatch) => {
    const notes = await loadNotes(uid);

    dispatch(setNotes(notes));
  }
}

export const startSaveNote = (note) => {
  return async(dispatch, getState) => {
    const {uid} = getState().auth;
    const noteToFirestore = {...note};
    
    noteToFirestore.date = new Date().getTime();
    delete noteToFirestore.id;

    try {
      db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore)
              .then( () => process.env.NODE_ENV !== "test" && Swal.fire('Updated', 'Your note has been updated!!', 'success'))
              .catch( (err) => process.env.NODE_ENV !== "test" && Swal.fire('Upss...', err.message, 'error '));
  
      dispatch(refreshNote(note.id, note));
    } catch (error) {
      throw new Error(error);
    }
  }
}

export const startUploadFile = (file) => {
  return async(dispatch, getState) => {
    const {active: activeNote} = getState().notes;

    if(process.env.NODE_ENV !== "test"){
      Swal.fire({
        title: 'Uploading',
        text: 'Please, wait a minute...',
        allowOutsideClick: false,
        showConfirmButton: false,
        willOpen: () => {
          Swal.showLoading();
        }
      })
    }

    try {
      const fileUrl = fileUpload(file);

      activeNote.imageUrl = fileUrl;
      activeNote.date = new Date().getTime();
      dispatch(startSaveNote(activeNote));
    } catch (error) {
      throw new Error(error);
    }
    Swal.close();
  }
}

export const startDeleteNote = (id) => {
  return async(dispatch, getState) => {
    const {uid} = getState().auth;
    await db.doc(`${uid}/journal/notes/${id}`).delete();

    dispatch(noteDelete(id));
  }
}

export const addNewNote = (id, note) => ({
  type: Types.notesAddNew,
  payload: {
    id,
    ...note
  }
})

export const activeNote = (id, note) => ({
  type: Types.notesActive,
  payload: {
    id,
    ...note
  }
});

export const setNotes = (notes) => ({
  type: Types.notesLoad,
  payload: notes
});

export const refreshNote = (id, note) => ({
  type: Types.notesUpdated,
  payload: {
    id,
    note
  }
})

export const noteDelete = (id) => ({
  type: Types.notesDeleted,
  payload: id
})

export const notesLogout = () => ({
  type: Types.notesLogoutCleaning
});