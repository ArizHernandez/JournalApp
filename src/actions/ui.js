import { Types } from "../types/types";

export const setError = (msgError) => ({
  type: Types.uiSetError,
  payload: msgError
})

export const removeError = () => ({
  type: Types.uiRemoveError
})

export const startLoading = () => ({
  type: Types.uiStartLoading
})

export const finishLoading = () => ({
  type: Types.uiFinishLoading
})