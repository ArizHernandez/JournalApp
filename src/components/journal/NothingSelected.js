import React from 'react'

export const NothingSelected = () => {
  return (
    <div className="nothing__main-content">
      <p>
        Select something
        <br />
          Or create an entry!
      </p>

      <i className="fa fa-meteor fa-4x mt-2"></i>
    </div>
  )
}
