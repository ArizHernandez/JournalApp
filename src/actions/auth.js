import Swal from 'sweetalert2';

import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { Types } from '../types/types';
import { notesLogout } from './notes';
import { startLoading, finishLoading, setError } from './ui';


export const startLoginEmailPassword = (email, password) => {
  return (dispatch) => {
    dispatch(startLoading());

    return firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({user}) => dispatch(login(user.uid, user.displayName)))
            .catch((err) => Swal.fire('Error!', err.message, 'error'))
            .finally(() => dispatch(finishLoading()))
  };
}

export const startRegisterEmailPasswordName = (email, password, name) => {
  return (dispatch) => {
    dispatch(startLoading());

    return firebase.auth().createUserWithEmailAndPassword(email, password)
            .then( async({user}) => {
              await user.updateProfile({displayName: name})

              dispatch(login(user.uid, user.displayName))
            })
            .catch((err) => Swal.fire('Error!', err.message, 'error'))
            .finally(() => dispatch(finishLoading()));
  }
}

export const startGoogleLogin = () => {
  return ((dispatch) => {
    firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({user}) => dispatch(login(user.uid, user.displayName)))
  });
}

export const startLogout = () => {
  return async(dispatch) => {
    await firebase.auth().signOut();
    dispatch(logout())
    dispatch(notesLogout());
  }
}

export const login = (uid, displayName) => ({
    type: Types.login,
    payload: {
      uid, 
      displayName
    }
})

export const logout = () => ({
  type: Types.logout
});