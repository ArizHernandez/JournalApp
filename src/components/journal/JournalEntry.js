import React from 'react'
import moment from 'moment';
import { useDispatch } from 'react-redux';
import { activeNote } from '../../actions/notes';

export const JournalEntry = ({id, body, date, imageUrl, title, i}) => {
  const dispatch = useDispatch();
  const noteDate = moment(date);

  const handleSelectNote = () => {

    dispatch(activeNote(id, {body, date, imageUrl, title}));
  }

  return (
    <div 
      className={`journal__entry animate__animated animate__fadeInLeft animate__delay-0${(i*10)/2}s`}
      onClick={handleSelectNote}
    >
      
      <div className="journal__entry-body-container">
        {
          (imageUrl !== "")
          && ( 
            <div 
              className="journal__entry-picture"
            >
              <img src={imageUrl} alt={title}/>
            </div>
          ) 
        }

        <div className="journal__entry-body">
          <p className="journal__entry-title">
            {title}
          </p>
          <p className="journal__entry-content">
            {body}
          </p>
        </div>

      </div>

      <div className="journal__entry-date-box">
        <span>{noteDate.format('dddd')}</span>
        <h4>{noteDate.format('Do')}</h4>
      </div>
    </div>
  )
}
