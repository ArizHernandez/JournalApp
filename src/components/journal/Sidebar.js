import React, { useState } from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { startLogout } from '../../actions/auth';
import { startNewNote } from '../../actions/notes';

import { JournalEtries } from './JournalEtries'

export const Sidebar = () => {
  const [actived, setActived] = useState(false);

  const dispatch = useDispatch();
  const {name} = useSelector(state => state.auth);

  
  const handleLogout = () => {
    dispatch(startLogout());
  }
  
  const handleAddNew = () => {
    dispatch(startNewNote());
  }

  return (
    <div className={`${actived && 'journal__sidebar-active'} journal__sidebar`}>
      <div 
        className="journal__menu-content"
        onClick={() => setActived(!actived)}
      >
        <div className={`${actived && 'journal__menu-item-closed'} journal__menu-item`}></div>
        <div className={`${actived && 'journal__menu-item-closed'} journal__menu-item`}></div>
      </div>
      
      <div className="journal__sidebar-navbar">
        <h3>
          <i className="fa fa-moon"></i>
          <span> {name}</span>
        </h3>

        <button 
          className={`btn journal__logout ${actived && "journal__logout-active"}`}
          onClick={handleLogout}
        >
          Logout
        </button>
      </div>

      <div 
        className="journal__new-entry"
        onClick={handleAddNew}
      >
        <i className="fa fa-calendar-plus fa-5x"></i>
        <p className="mt-2">New entry</p>
      </div>

      <JournalEtries />
    </div>
  )
}
