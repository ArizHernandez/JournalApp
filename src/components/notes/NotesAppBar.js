import moment from 'moment';
import React from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { startDeleteNote, startSaveNote, startUploadFile } from '../../actions/notes';

export const NotesAppBar = ({date, id}) => {
  const dispatch = useDispatch();
  const {active} = useSelector(state => state.notes)
  const newDate = moment(date).format("D [de] MMMM [de] YYYY");

  const handleSave = () => {
    dispatch(startSaveNote(active))
  }

  const handleUploadPicture = () => {
    document.querySelector('#fileSelector').click();
  }
  
  const handleFireChange = (e) => {
    const file = e.target.files[0];
    
    if(file){
      dispatch(startUploadFile(file));
      document.querySelector('#fileSelector').value = "";
    }
  }

  const handleDelete = () => {
    dispatch(startDeleteNote(id));
  }

  return (
    <div className="notes__appbar">
      <span>{newDate}</span>

      <input  
        id="fileSelector"
        type="file"
        style={{display:"none"}}
        onChange={handleFireChange}
      />

      <div>
        <button 
          className="btn animate__animated animate__fadeInDown"
          onClick={handleUploadPicture}
        >
          Picture
        </button>

        <button 
          className="btn animate__animated animate__fadeInDown animate__delay-010s"
          onClick={handleSave}  
        >
          save
        </button>

        <button
          className="btn btn-danger animate__animated animate__fadeInDown animate__delay-020s"
          onClick={handleDelete}
        >
          Delete
        </button>
      </div>
      
    </div>
  )
}
