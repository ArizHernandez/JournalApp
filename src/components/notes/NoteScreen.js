import React, { useEffect, useRef } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { activeNote } from '../../actions/notes';

import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar'

export const NoteScreen = () => {
  const dispatch = useDispatch();
  const {active: note} = useSelector(state => state.notes);
  const [formValue, handleInputChange, reset] = useForm(note);
  const {title, body} = formValue;

  const activeId = useRef(note.id);

  useEffect(() => {
    if(note.id !== activeId.current){
      reset(note);
      activeId.current = note.id;
    }
  }, [note, reset]);

  useEffect(() => {
    dispatch(activeNote(formValue.id, {...formValue}));
  }, [formValue, dispatch])

  return (
    <div className="notes__main-content animate__animated animate__fadeIn">
      
      <NotesAppBar {...formValue}/> 
      
      <div className="notes__content">
        <input
          type="text"
          placeholder="Write a new title"
          className="notes__title-input mb-2 animate__animated animate__fadeIn animate__fast"
          autoComplete="off"
          name="title"
          value={title}
          onChange={handleInputChange}
        />

        <textarea
          placeholder="What happend today?"
          className="notes__text-area mb-2"
          name="body"
          value={body}
          onChange={handleInputChange}
        />

        {
          (note.imageUrl)
          && (
            <div className="notes__image">
              <img 
                src={note.imageUrl}
                alt="journal-canguro"
              />
            </div>
          )
        }

      </div>

    </div>
  )
}
