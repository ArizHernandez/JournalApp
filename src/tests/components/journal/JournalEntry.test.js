import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { JournalEntry } from '../../../components/journal/JournalEntry';
import { activeNote } from '../../../actions/notes';

jest.mock('../../../actions/notes', () => ({
  activeNote: jest.fn()
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initialState = {};

const store = mockStore(initialState);
store.dispatch = jest.fn();

describe('Pruebas en el <JournalEntry />', () => {
  const wrapper = mount(
    <Provider store={store}>
      <JournalEntry
        id="1"
        body="test body"
        title="test title"
        date={123456}
        imageUrl="google.com"
        i={1}
      />
    </Provider>
  )

  test('Debe de cargar correctamente', () => {
    expect(wrapper).toMatchSnapshot();
  })
  
  test('Debe de disparar el activeNote', () => {
    wrapper.find('.journal__entry').prop('onClick')();

    expect(activeNote).toHaveBeenCalled();
    expect(activeNote).toHaveBeenCalledWith("1",{
      body: "test body", 
      date: 123456, 
      imageUrl: "google.com", 
      title: "test title"
    });
  })
})