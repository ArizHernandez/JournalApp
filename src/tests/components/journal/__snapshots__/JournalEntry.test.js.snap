// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <JournalEntry /> Debe de cargar correctamente 1`] = `
<div
  className="journal__entry animate__animated animate__fadeInLeft animate__delay-05s"
  onClick={[Function]}
>
  <div
    className="journal__entry-body-container"
  >
    <div
      className="journal__entry-picture"
    >
      <img
        alt="test title"
        src="google.com"
      />
    </div>
    <div
      className="journal__entry-body"
    >
      <p
        className="journal__entry-title"
      >
        test title
      </p>
      <p
        className="journal__entry-content"
      >
        test body
      </p>
    </div>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Wednesday
    </span>
    <h4>
      31st
    </h4>
  </div>
</div>
`;
